[CCA][Memory Snapshot] Create TraceEntryRingBuffer class for alloc_trace logic

aaronenyeshi opened the pull request:
Summary:
Move the alloc_trace logic into a separate class, to reduce risk of deadlocks when mixing with CCA's lock. Switch to an std::mutex instead of std::recursive_mutex.

Let's us re-use the logic in TraceEntryRingBuffer class for later diffs.

Test Plan: CI, resnet run, and FBR model.

Differential Revision: D59690408

Pulled By: aaronenyeshi

davidberard98 commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -792,6 +792,64 @@ cudaError_t cudaMallocMaybeCapturing(void** p, size_t size) {
+  }
+  
+  void setMaxEntries(size_t size) {
+    alloc_trace_max_entries_ = std::max(size_t(1), size);;
```
I think we should lock here too?

aaronenyeshi replied:
Added!

davidberard98 reviewed (approved):
one comment about adding a lock, and also take a look at the OSS linter - otherwise LGTM!

aaronenyeshi commented:
Fixed the lock comment, and lintrunner / clang-tidy warnings. CI is re-running.

aaronenyeshi committed:
[CCA][Memory Snapshot] Create TraceEntryRingBuffer class for alloc_trace logic (#130741)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/130741

Move the alloc_trace logic into a separate class, to reduce risk of deadlocks when mixing with CCA's lock. Switch to an std::mutex instead of std::recursive_mutex.

Let's us re-use the logic in TraceEntryRingBuffer class for later diffs.

Test Plan: CI, resnet run, and FBR model.

Reviewed By: davidberard98

Differential Revision: D59690408

Pulled By: aaronenyeshi

aaronenyeshi commented:
All CI signals passed/flaky except these two:
```
[pull / linux-focal-cuda12.1-py3.10-gcc9-sm86 / test (default, 1, 5, linux.g5.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/130741#27478718480) ([gh](https://github.com/pytorch/pytorch/actions/runs/9944895747/job/27478718480))
inductor/test_cudagraph_trees.py::CudaGraphTreeTests::test_fallback_to_eager_if_recompiling_too_many_times
[trunk / linux-focal-cuda12.4-py3.10-gcc9-sm86 / test (default, 1, 5, linux.g5.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/130741#27477926813) ([gh](https://github.com/pytorch/pytorch/actions/runs/9944896710/job/27477926813))
inductor/test_cudagraph_trees.py::CudaGraphTreeTests::test_fallback_to_eager_if_recompiling_too_many_times
```

However, it looks like these two were disabled 20 hours ago so I assume they are broken.

aaronenyeshi commented:
@pytorchbot merge -f "Phabricator Diff has merged, these failed signals are pre-existing"
